### PR TITLE
Add remote search to `brew cask search`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Adds functionality to `brew cask search` to search remote Taps.

Closes https://github.com/caskroom/homebrew-cask/issues/16732.

### CODE REVIEW

File: Library/Homebrew/cask/lib/hbc/cli/search.rb [hunk 1]
```diff
@@ -13,6 +13,22 @@ def self.extract_regexp(string)
...
+        matches.map do |match|
+          dirname, filename = File.split(match["path"])
+          next unless dirname == "Casks"
+          user = match["repository"]["owner"]["login"]
+          repo = match["repository"]["name"]
+          tap = Tap.fetch(user, repo)
+          next if tap.installed?
+          basename = File.basename(filename, ".rb")
+          "#{tap.name}/#{basename}"
+        end.compact
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: This logic is very similar and looks like pretty much all except the specific dirname could be shared.

File: Library/Homebrew/cmd/search.rb [hunk 2]
```diff
@@ -102,20 +102,17 @@ def query_regexp(query)
...
-        user = object["repository"]["owner"]["login"]
-        user = user.downcase if user == "Homebrew"
-        repo = object["repository"]["name"].sub(/^homebrew-/, "")
-        tap = Tap.fetch user, repo
-        next if tap.installed?
-        basename = File.basename(filename, ".rb")
-        "#{user}/#{repo}/#{basename}"
-      end.compact
-    end
+    matches = GitHub.search_code(users: ["Homebrew", 'caskroom'], filename: query)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]: Use `"` consistently

File: Library/Homebrew/utils/github.rb [hunk 3]
```diff
@@ -228,13 +228,23 @@ def raise_api_error(output, errors, http_code, headers, scopes)
...
     uri.query = build_query_string(query, qualifiers)
     open(uri) { |json| json["items"] }
   end
 
   def repository(user, repo)
-    open(URI.parse("https://api.github.com/repos/#{user}/#{repo}")) { |j| j }
+    open(URI.parse("#{API_URL}/repos/#{user}/#{repo}")) { |j| j }
+  end
+
+  def search_code(filename: nil, filenames: nil, user: nil, users: nil, &block)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]: Kill `filenames` (it's never used) and `user` in favour of passing an array. These simplify both the parameters and code below.

File: Library/Homebrew/utils/github.rb [hunk 4]
```diff
@@ -228,13 +228,23 @@ def raise_api_error(output, errors, http_code, headers, scopes)
...
+    filenames = [*filename, *filenames].map { |f| "filename:#{f}" }
+    users = [*user, *users].map { |u| "user:#{u}" }
+
+    uri = URI.parse("#{API_URL}/search/code")
+    uri.query = "q=".concat(uri_escape((users + filenames).join(" ")))
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]: Use `#{}` rather than `concat`

File: Library/Homebrew/cmd/search.rb [hunk 5]
```diff
@@ -101,21 +101,15 @@ def query_regexp(query)
...
-        user = object["repository"]["owner"]["login"]
-        user = user.downcase if user == "Homebrew"
-        repo = object["repository"]["name"].sub(/^homebrew-/, "")
-        tap = Tap.fetch user, repo
-        next if tap.installed?
-        basename = File.basename(filename, ".rb")
-        "#{user}/#{repo}/#{basename}"
-      end.compact
-    end
+    valid_dirnames = ["Formula", "HomebrewFormula", "Casks", "."].freeze
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]: Are you sure you can drop `""` here? Have you verified this with formulae in the root e.g. homebrew/science?

File: Library/Homebrew/utils/github.rb [hunk 6]
```diff
@@ -228,13 +228,19 @@ def raise_api_error(output, errors, http_code, headers, scopes)
...
     uri.query = build_query_string(query, qualifiers)
     open(uri) { |json| json["items"] }
   end
 
   def repository(user, repo)
-    open(URI.parse("https://api.github.com/repos/#{user}/#{repo}")) { |j| j }
+    open(URI.parse("#{API_URL}/repos/#{user}/#{repo}"))
+  end
+
+  def search_code(*params)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]: Maybe `(*args)` is more idiomatic? No strong preferences though.

File: Library/Homebrew/cmd/search.rb (hunk 5 above)
Comment (reitermarkus) [inline Library/Homebrew/cmd/search.rb]: Yes, since `File.split("formula.rb")` returns `[".", "formula.rb"]`.

Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]: 🆒. :shipit: if homebrew/science works (`brew search octave` with it untapped)

File: Library/Homebrew/utils/github.rb (hunk 6 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/github.rb]: Yeah, I thought of these as URL parameters, hence `params`.

Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]: 🆒

**Pull request merged**